### PR TITLE
Added EarthSurfaceWater tutorial

### PR DESCRIPTION
**Pull request by cordmaur**
Here is the tutorial as demanded by James.

### CODE REVIEW

**Comment by ashnair1** [comment]
Thanks a lot for the tutorial @cordmaur. 

Could you change the name of the tutorial to `earthsurfacewater.ipynb` and add the tutorial to `docs/index.rst`?

**Comment by adamjstewart** [comment]
I would actually name it `earth_surface_water.ipynb`

P.S. I'll try to review this in more detail when I get a chance. We're currently rushing to finish a new TorchGeo release in the next week or two so we have a long list of PRs to merge before then.

**Comment by adamjstewart** [comment]
@cordmaur is this PR still a work in progress? If so you need to accept the CLA and make a few changes to get the tests working.

**Comment by cordmaur** [comment]
@microsoft-github-policy-service agree

**Comment by cordmaur** [comment]
Hi @adamjstewart. I finished updating the tutorial. I could get all tests passing, except for the `docs`. I added it to the `docs/index.rst`, but following the errors in my local `make html` there are issues in `demo.rst`.

**Comment by adamjstewart** [comment]
I think if you rename one of the images from `[Image]` to something else, it should be fine. It's just complaining that the alt text is the same for multiple images.

The time our notebook tests take to run increased from 9 min to 29 min. We'll have to find a way to make the tests run faster (either download less data or monkeypatch to run for fewer epochs/steps).

**Comment by cordmaur** [comment]
Hi, @adamjstewart, all tests are passing. 
Notebook checks are running on 10min...

**Comment by adamjstewart** [comment]
Thanks, will give this a detailed review as soon as I get a chance!

File: docs/tutorials/earth_surface_water.ipynb [hunk 1]
```diff
@@ -0,0 +1,1339 @@
...
+    "<a href=\"https://colab.research.google.com/gist/cordmaur/50db0a00792c090e5a5cfcebede58b9e/copy-of-earthsurfacewater_torchgeo_tutorial.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "MAR190Aszv8r"
+   },
+   "source": [
+    "# Earth Water Surface Tutorial\n",
```
Comment (adamjstewart) [inline docs/tutorials/earth_surface_water.ipynb:19]:
```suggestion
    "# Earth Surface Water\n",
```

File: docs/tutorials/earth_surface_water.ipynb [hunk 2]
```diff
@@ -0,0 +1,1339 @@
...
...dcHPeGzv7adH7r/8/Pd2yaHY6OPT7/c7BiDNL5f/175HzU3Bz8/Roc4CDseM+ACUfHGv+cn6tnQOjcV/BRx3cYnZM3bVcmhh8Pfz1vS2wFp2P53/3zdQ7O+XqGSlPLtfu57j9r1GRdf2RMBIzzElUTEDHFFFNMMcVvO37NQIRE28oXXpaFXF6mBMGSum6IosjJnDREUU3btiOAQZNoc0qIfWeCaIN6k7dxpw5woou+68cn42pIBwpGjE2szw2f5F+vk6rAhBhFNS456E7Jiu+MEbpw24aUpXQalGV9AkW0kJplCVEUI/4XA1XVOF8GT7MMgsF1WGuXlCbCEQJIyDWpxqV0A4kXhTGScEF82p+wIJRSbAjDEGtbkqQmTQ+kaeS6kyxxHBDHWqSPyLKMxWLOarViuVyRJAfu7go2myWHw0dcXCzpuoE4zvnpT39BUfR8+mnJs2fvs1hAksQEQe1+SoIgQ3wbrjEmIE2h60LquscYiKIUTx3HjYfIFQkFO3DGwa0rZlvCMCUMB6IoJo5z0nRGHM+Yzeak6YwgwLEUOg6HNxyPD8TxQBQlDgw5UFUJUTRDHw1dcAxDgyys3n5ktDPO2uT0WhiGvPfeR/zJn7xgv3/gxYt7qmrg4WFDXYv599/+7b/nk08+48/+7M/59//+Bd99dwQCBzwFtG1LXffOvK/nxYt7/vzPf8GXX74BYpqmoywLiqKkrmuGYSBJElarLa9e7Xnx4g1/+Iff8vnnv8v19YfE8TUCLOhc/T7xrsXp+HeDsnQ8MJiPthmQRWKD10+Vxd1qFZKmGbNZzpMnV1xefkuWveDhYUdVVRwONYdDCwROAkk0mK21Trqpc5JKYgwPOXEsMlvHY8Ph0HA4iLxSmhry3BDHOUkiwEHTxHSddMa1be+eb1loGxOewAr5HAjcNr373GppW28OLwteWRCIsXvIxcUlWbYiSQzWTkn+FFNMMcUUU/wm4rxhYnDFyPNiphabtTnCF/4ea6ifMycB1DNCzJ3V1FmbGDRHB9+hHY4KvAPjNc25vI2+UcCSvpc1inhLBU7+ST0kZB+SczZubSSeaP5fDz54Hwq95t4BAOox17ji5GmERoVgLbB2jj1QuaYN68AIvVbdVrvBxdBbQIOeKOqJIgiCiGGIEKZK4PbbnQAfAY4ESFC5TTlvvWfaWKZAD6iclRbIvbH4/4
```
Comment (adamjstewart) [inline docs/tutorials/earth_surface_water.ipynb:621]: Let's use a math expression instead, this is very blurry

Comment (adamjstewart) [inline docs/tutorials/earth_surface_water.ipynb:672]: Can we use `kornia.augmentation.Normalize` instead? Looks identical

**Review by adamjstewart — CHANGES_REQUESTED** [review]
First round of reviews. Overall, looks great!

FYI: we are presenting a pre-conference workshop (basically a tutorial) on TorchGeo at AGU this weekend. I would love to finish and merge this PR soon so we can include it in the talk. If you don't mind, I can make direct changes to your PR branch tomorrow night after I finish my other sections. But anything you can fix before then would save me a ton of time.

File: docs/tutorials/earth_surface_water.ipynb [hunk 3]
```diff
@@ -0,0 +1,1339 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "view-in-github"
+   },
+   "source": [
+    "<a href=\"https://colab.research.google.com/gist/cordmaur/50db0a00792c090e5a5cfcebede58b9e/copy-of-earthsurfacewater_torchgeo_tutorial.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:10]: Removed

File: docs/tutorials/earth_surface_water.ipynb [hunk 4]
```diff
@@ -0,0 +1,1339 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "MAR190Aszv8r"
+   },
+   "source": [
+    "# Earth Water Surface Tutorial\n",
+    "\n",
+    "Author: Mauricio Cordeiro<br>\n",
+    "LinkedIn: https://www.linkedin.com/in/cordmaur/<br>\n",
+    "Medium: https://cordmaur.medium.com/\n",
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:23]: Removed the links.

File: docs/tutorials/earth_surface_water.ipynb [hunk 5]
```diff
@@ -0,0 +1,1339 @@
...
+    "The objective of this tutorial is to go through the Earth Water Surface dataset and cover the following topics:\n",
+    "* Creating RasterDatasets, DataLoaders and Samplers for images and masks;\n",
+    "* Intersection Dataset;\n",
+    "* Normalizing the data;\n",
+    "* Creating spectral indices;\n",
+    "* Creating the segmentation model (DeepLabV3);\n",
+    "* Loss function and metrics; and\n",
+    "* Training loop.\n",
+    "\n",
+    "> This tutorial is also available as a 3 parts Medium story: <br>https://medium.com/towards-data-science/artificial-intelligence-for-geospatial-analysis-with-pytorchs-torchgeo-part-1-52d17e409f09"
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:51]: Moved to the `Additional Reading` secion (at the bottom)

File: docs/tutorials/earth_surface_water.ipynb (hunk 2 above)
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:672]: Indeed... just had to "workaround" to avoid normalizing the spectral indices... 
```
# normalize = MyNormalize(*calc_statistics(train_imgs))
mean, std = calc_statistics(train_imgs)

# Please, note that we will create spectral indices using the raw (non-normalized) data. Then, when normalizing, the sensors will have more channels (the indices) that should not be normalized.
# To solve this, we will add the indices to the 0's to the mean vector and 1's to the std vectors

mean = np.concat([mean, [0, 0, 0]])
std = np.concat([std, [1, 1, 1]])

norm = K.Normalize(mean=mean, std=std)

tfms = torch.nn.Sequential(
    indices.AppendNDWI(index_green=1, index_nir=3),
    indices.AppendNDWI(index_green=1, index_nir=5),
    indices.AppendNDVI(index_nir=3, index_red=2),
    norm,
)
```

[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:1223]: I can't see where you suggested this....

Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:621]: Done.

File: docs/tutorials/earth_surface_water.ipynb [hunk 6]
```diff
@@ -0,0 +1,1339 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# this line is to avoid warnings from Rasterio to be printed on screen\n",
+    "# the warning is explained here: https://stackoverflow.com/questions/74089170/suppress-rasterio-warning-warning-1-tiffreaddirectory\n",
+    "logger = logging.getLogger('rasterio')\n",
+    "logger.setLevel(logging.ERROR)\n",
+    "\n",
+    "warnings.filterwarnings('ignore', message='TIFFReadDirectory')\n",
+    "warnings.filterwarnings('ignore', module='osgeo')"
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:153]: WIth the images corrected, I could remove the `warnings` dependency.

File: .github/workflows/tutorials.yaml [hunk 7]
```diff
@@ -38,7 +38,7 @@ jobs:
       - name: List pip dependencies
         run: pip list
       - name: Run notebook checks
-        run: pytest --nbmake --durations=10 docs/tutorials
+        run: pytest --nbmake --durations=10 docs/tutorials --nbmake-timeout=3000
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply .github/workflows/tutorials.yaml]: Removed the timeout limit and it's still passing.

File: docs/tutorials/earth_surface_water.ipynb [hunk 8]
```diff
@@ -0,0 +1,1339 @@
...
+   "metadata": {},
+   "source": [
+    "### Data correction"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The Tiffs in this dataset are raising an warning because the number of values by pixel does not match the Photometric Tiff tag value. This problem is explained <a href=\"https://gis.stackexchange.com/questions/350084/warning1-tiffreaddirectorysum-of-photometric-type-related-color-channels\">here</a> and <a href=\"https://github.com/geosolutions-it/imageio-ext/issues/164\">here</a>.\n",
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb]: I don't have rights to create a dataset on huggingface, so I corrected the dataset and hosted it under cordmaur:
'https://huggingface.co/datasets/cordmaur/earth_surface_water/resolve/main/earth_surface_water.zip'

Feel free to move it to torchgeo.

File: docs/tutorials/earth_surface_water.ipynb [hunk 9]
```diff
@@ -0,0 +1,1339 @@
...
+    "\n",
+    "cd /tmp/dset-s2/val_scene\n",
+    "for file in $(ls *.tif); do\n",
+    "    tiffset -s 338 6 0 0 0 0 0 0 $file 2> /tmp/warnings.log\n",
+    "done\n",
+    "\n",
+    "cd /tmp/dset-s2/tra_scene\n",
+    "for file in $(ls *.tif); do\n",
+    "    tiffset -s 338 6 0 0 0 0 0 0 $file 2> /tmp/warnings.log\n",
+    "done"
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb]: Removed.

File: docs/tutorials/earth_surface_water.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb]: I don't see exactly where you commented this... maybe it's to replace the OA and IOU?

File: docs/tutorials/earth_surface_water.ipynb [hunk 10]
```diff
@@ -0,0 +1,1339 @@
...
+      "\n",
+      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m24.3.1\u001b[0m\n",
+      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
+      "Note: you may need to restart the kernel to use updated packages.\n"
+     ]
+    }
+   ],
+   "source": [
+    "# %pip install rasterio -q\n",
+    "# %pip install torchgeo -q\n",
```
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb]: Removed the comments. Scikit is being used only for the IoU metrics... maybe I can replace it to torchmetrics. Will work on this...

File: docs/tutorials/earth_surface_water.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb]: Moved to Additional Reading.

File: docs/tutorials/earth_surface_water.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb]: Ok. Removed.

File: docs/tutorials/earth_surface_water.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (cordmaur) [reply docs/tutorials/earth_surface_water.ipynb:1192]: How can we do this if we don't have enough time to "train" the model during commit?

**Comment by cordmaur** [comment]
Hi @adamjstewart . Solved most of the requests. 
Corrected dataset is being hosted on my hugging face because I didnt have the rights to torchgeo repository. You can move it or grant me the rights.

I think it's only missing the `scikit` dependency, which can be resolved using `torchmetrics` and the final evaluation... Feel free to change these directly. I will only hava time to work on those tomorrow...

File: docs/tutorials/earth_surface_water.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply docs/tutorials/earth_surface_water.ipynb:1114]: Yes, OA and IoU.

[the remark this answers is not in the document]
Comment (adamjstewart) [reply docs/tutorials/earth_surface_water.ipynb:1192]: We'll likely need to use a pre-trained model end transfer learning. We can test for 1 epoch during CI and many epochs for the actual tutorial. See my Introduction to PyTorch tutorial PR for details.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Thanks for the contribution, glad we finally got this in!
